### PR TITLE
Fix return type value from info field

### PR DESCRIPTION
**Pull request by la4de**
This fixes the bug where return type was some times resolved incorrectly.

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release fixes the return type value from info argument of resolver.

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release fixes the return type value from info argument of resolver.

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release fixes the return type value from info argument of resolver.

File: tests/schema/test_info.py [hunk 1]
```diff
@@ -62,3 +64,38 @@ def hello(self, info: Info[str, str]) -> Result:
...
+
+
+def test_return_type():
+    def resolver_factory(return_type, return_value):
+        def resolver(info) -> return_type:
+            # check return type
+            assert info.return_type == return_type
+            return return_value
+
+        return resolver
```
Comment (patrick91) [inline tests/schema/test_info.py:76]: could we use pytest parametrization for this? 😊

Comment (la4de) [reply tests/schema/test_info.py:76]: Nice idea, but I'm not sure if it'd be easy task. I added different type of resolvers too. some return types are inherited from resolver, some from field

Comment (patrick91) [reply tests/schema/test_info.py:76]: we can split the test in two 😊

Comment (la4de) [reply tests/schema/test_info.py:76]: that's even better idea =)

File: strawberry/field.py [hunk 2]
```diff
@@ -250,7 +255,7 @@ def _strawberry_info_from_graphql(info: GraphQLResolveInfo) -> Info:
                 context=info.context,
                 root_value=info.root_value,
                 variable_values=info.variable_values,
-                return_type=self.type,
+                return_type=self._return_type,
```
Comment (BryceBeagle) [inline strawberry/field.py:258]: Can we just use 
```suggestion
                return_type=self.type or self.base_resolver.type,
```
and not store the extra variable?

**Comment by la4de** [comment]
I have a question. Is this valid case to leave return type from resolver function and define it only in field? I'd expect that this would return type `int` and not `None`.

```python
        def resolver(info):
            assert info.return_type == expected
            return return_value

        @strawberry.type
        class Query:
            field: int = strawberry.field(resolver=resolver)
 ```

**Comment by patrick91** [comment]
> I have a question. Is this valid case to leave return type from resolver function and define it only in field? I'd expect that this would return type `int` and not `None`.
> 
> ```python
>         def resolver(info):
>             assert info.return_type == expected
>             return return_value
> 
>         @strawberry.type
>         class Query:
>             field: int = strawberry.field(resolver=resolver)
> ```

I think we briefly discussed this, but never decided if we should enforce types of resolver defined with `resolver=resolver`

Internally we could change the return type (at least on our StrawberryResolver object)

**Comment by BryceBeagle** [comment]
Yeah that should probably do one of two things:
1. Use the field type because the resolver is not annotated
2. Raise an exception because the resolver's (implicit `None`) return annotation does not match the field's type

My vote would be for #2

**Comment by patrick91** [comment]
> Yeah that should probably do one of two things:
> 
> 1. Use the field type because the resolver is not annotated
> 2. Raise an exception because the resolver's (implicit `None`) return annotation does not match the field's type
> 
> My vote would be for #2

#2 would probably be a big breaking change, I think we should only throw if there's a type that doesn't match, at least for now :)

**Comment by BryceBeagle** [comment]
> #⁠2 would probably be a big breaking change

I guess this could be another idea for the "things configurable on the Schema" bucket

**Comment by patrick91** [comment]
> > #⁠2 would probably be a big breaking change
> 
> I guess this could be another idea for the "things configurable on the Schema" bucket

gotta love feature flags :D

**Comment by la4de** [comment]
Many tests start to fail, if I access `self.base_resolver.type` in `_strawberry_info_from_graphql()`. In other other words, following change
```diff
--- a/strawberry/field.py
+++ b/strawberry/field.py
@@ -244,6 +244,7 @@ class StrawberryField(dataclasses.Field):
             return result

         def _strawberry_info_from_graphql(info: GraphQLResolveInfo) -> Info:
+            self.base_resolver.type
             return Info(
                 field_name=info.field_name,
                 field_nodes=info.field_nodes,
```

causes this
```
> pytest
...
=== 83 failed, 354 passed, 2 xfailed in 40.93s ===
```

**Review by la4de — COMMENTED** [review]
> Many tests start to fail, if I access `self.base_resolver.type` in `_strawberry_info_from_graphql()`. In other other words, following change
> 
> ```diff
> --- a/strawberry/field.py
> +++ b/strawberry/field.py
> @@ -244,6 +244,7 @@ class StrawberryField(dataclasses.Field):
>              return result
> 
>          def _strawberry_info_from_graphql(info: GraphQLResolveInfo) -> Info:
> +            self.base_resolver.type
>              return Info(
>                  field_name=info.field_name,
>                  field_nodes=info.field_nodes,
> ```
> 
> causes this
> 
> ```
> > pytest
> ...
> === 83 failed, 354 passed, 2 xfailed in 40.93s ===
> ```

I finally found the reason. `base_resolver` is sometimes `None` and `None` does not have `type` attribute.

It would be very nice to get full back traces from errors in unit tests. Now GraphQL exception capturing making all this invisible :)

File: tests/schema/test_info.py (hunk 1 above)
Comment (la4de) [reply tests/schema/test_info.py]: Done

File: strawberry/field.py (hunk 2 above)
Comment (la4de) [reply strawberry/field.py]: After playing with the code and debugging multiple weird issues I chose another approach. Please take a look :)

**Comment by la4de** [comment]
@patrick91 , @BryceBeagle, I ended up implementing getter for return type. At least functional tests are happy now. Unfortunately mypy wasn't so happy. I'd need some help from you now to resolve that.

I'm running out of ideas how to solve typing issues. Could you please take a look, what's wrong there. Please also review the code. Thanks.

File: strawberry/field.py [hunk 3]
```diff
@@ -274,6 +274,15 @@ async def await_result(result):
         _resolver._is_default = not self.base_resolver  # type: ignore
         return _resolver
 
+    def _get_return_type(self):
+        if self.is_list:
+            type_ = List[self.child._get_return_type()]
```
Comment (patrick91) [inline strawberry/field.py:279]: worth nothing that this in future will change to StrawberryList 😊

**Review by patrick91 — APPROVED** [review]
Looks good to me! Let's double check with @BryceBeagle before merging 😊

File: strawberry/field.py (hunk 3 above)
Comment (BryceBeagle) [reply strawberry/field.py:279]: Or potentially `StrawberryType` with `.is_list == True`. I'm still waffling on the implementation...

**Review by BryceBeagle — APPROVED** [review]
This looks to me like a good enough fix until `StrawberryType` is implemented

File: strawberry/field.py (hunk 3 above)
Comment (patrick91) [reply strawberry/field.py:279]: ok! no rush, I think this will be only used internally for now 😊

**Pull request merged**